tests/scattering2d/test_jax_scattering2d.py

```diff
@@ -0,0 +1,256 @@
+    def data(self, request):
+        real_filter = request.param
+        x = (np.random.randn(100, 128, 128)
+             + 1J * np.random.randn(100, 128, 128))
+        filt = (np.random.randn(128, 128)
+                + 1J * np.random.randn(128, 128))
+        y = (np.random.randn(100, 128, 128)
+             + 1J * np.random.randn(100, 128, 128))
+
+        if real_filter:
+            filt = np.real(filt)
+
+        y = x * filt
+
+        return x, filt, y
+
+    def test_cdgmm_forward(self, data):
+        x, filt, y = data
+
+        z = backend.cdgmm(x, filt)
+
+        assert np.allclose(y, z)
+
+def test_cdgmm_exceptions():
+    with pytest.raises(TypeError) as record:
+        backend.cdgmm(np.empty((3, 4, 5)).astype(np.float64),
+                        np.empty((4, 5)).astype(np.complex128))
+    assert 'first input must be complex' in record.value.args[0]
+
+    with pytest.raises(TypeError) as record:
+        backend.cdgmm(np.empty((3, 4, 5)).astype(np.complex128),
+                        np.empty((4, 5)).astype(np.int64))
+    assert 'second input must be complex or real' in record.value.args[0]
+
+    with pytest.raises(RuntimeError) as record:
+        backend.cdgmm(np.empty((3, 4, 5)).astype(np.complex128),
+                        np.empty((4, 6)).astype(np.complex128))
+    assert 'not compatible for multiplication' in record.value.args[0]
+
+
+def test_fft():
+    np.random.seed(9161341)
+    x = np.random.randn(2, 2) + 1J * np.random.randn(2, 2)
+
+    y = np.array([[x[0, 0] + x[0, 1] + x[1, 0] + x[1, 1],
+                    x[0, 0] - x[0, 1] + x[1, 0] - x[1, 1]],
+                    [x[0, 0] + x[0, 1] - x[1, 0] - x[1, 1],
+                    x[0, 0] - x[0, 1] - x[1, 0] + x[1, 1]]])
+
+
+    z = backend.ifft(x)
+
+    z = z * 4
+
+    assert np.allclose(y, z)
+
+    z = backend.irfft(x)
+
+    z = z * 4
+
+    assert not np.iscomplexobj(z)
+    assert np.allclose(y.real, z)
+
+    x = np.random.randn(2, 2)
+
+    y = np.array([[x[0, 0] + x[0, 1] + x[1, 0] + x[1, 1],
+                    x[0, 0] - x[0, 1] + x[1, 0] - x[1, 1]],
+                    [x[0, 0] + x[0, 1] - x[1, 0] - x[1, 1],
+                    x[0, 0] - x[0, 1] - x[1, 0] + x[1, 1]]])
+
+    z = backend.rfft(x)
+    assert np.allclose(y.real, z)
+
+
+def test_concatenate():
+    x = np.random.randn(3, 6, 6) + 1J * np.random.randn(3, 6, 6)
+    y = np.random.randn(3, 6, 6) + 1J * np.random.randn(3, 6, 6)
+    z = np.random.randn(3, 6, 6) + 1J * np.random.randn(3, 6, 6)
+
+    w = backend.concatenate((x, y, z))
+
+    assert w.shape == (x.shape[0],) + (3,) + (x.shape[-2:])
+    assert np.allclose(w[:, 0, ...], x)
+    assert np.allclose(w[:, 1, ...], y)
+    assert np.allclose(w[:, 2, ...], z)
+
+
+def test_Scattering2D():
+    test_data_dir = os.path.dirname(__file__)
+    data = None
+    with open(os.path.join(test_data_dir, 'test_data_2d.npz'), 'rb') as f:
+        buffer = io.BytesIO(f.read())
+        data = np.load(buffer)
+
+    x = data['x']
```

Comment:
Shouldn't we be casting to JNP arrays or doing `device_put` or something here (like in 1D)?

Reply:
we should also be parameterizing Jax with cpu and gpu device in that instance then...

Reply:
I agree.